osu.Game/Utils/ModUtils.cs

```diff
@@ -295,43 +296,56 @@ public static double CalculateRateWithMods(IEnumerable<Mod> mods)
+        /// <param name="isRequired">Whether the mod is intended as a "required" (room-global) mod.</param>
+        /// <param name="isFreestyle">Whether freestyle is enabled for the playlist item.</param>
+        /// <seealso href="https://github.com/ppy/osu-web/blob/40936b514c6485b874f6c6496d55d9e8b1b88fd4/app/Singletons/Mods.php#L95-L113">Related osu!web function.</seealso>
+        public static bool IsValidModForMatch(Mod mod, MatchType matchType, bool isRequired, bool isFreestyle)
         {
             if (mod.Type == ModType.System || !mod.UserPlayable || !mod.HasImplementation)
                 return false;
 
-            switch (type)
+            if (isFreestyle && isRequired && !mod.ValidForFreestyleAsRequiredMod)
+                return false;
+
+            switch (matchType)
             {
                 case MatchType.Playlists:
                     return true;
 
                 default:
-                    return mod.ValidForMultiplayer;
+                    return isRequired ? mod.ValidForMultiplayer : mod.ValidForMultiplayerAsFreeMod;
             }
         }
 
         /// <summary>
-        /// Determines whether a mod can be applied as a free mod to playlist items in the given match type.
+        /// Given an online listing of mods and the user's preferred ruleset, gathers the mods which are selectable as free mods by the current user.
         /// </summary>
-        /// <param name="mod">The mod to test.</param>
-        /// <param name="type">The match type.</param>
-        public static bool IsValidFreeModForMatchType(Mod mod, MatchType type)
+        /// <param name="matchType">The type of match being played.</param>
+        /// <param name="requiredMods">The required mods for the playlist item.</param>
+        /// <param name="allowedMods">The allowed mods for the playlist item.</param>
+        /// <param name="freestyle">Whether freestyle is enabled for the playlist item.</param>
+        /// <param name="userRuleset">The user's preferred ruleset, which may differ from the playlist item's selection on freestyle playlist items.</param>
+        public static Mod[] ListUserSelectableFreeMods(MatchType matchType, IEnumerable<APIMod> requiredMods, IEnumerable<APIMod> allowedMods, bool freestyle, Ruleset userRuleset)
         {
-            if (mod.Type == ModType.System || !mod.UserPlayable || !mod.HasImplementation)
-                return false;
-
-            switch (type)
+            if (freestyle)
             {
-                case MatchType.Playlists:
-                    return true;
-
-                default:
-                    return mod.ValidForMultiplayerAsFreeMod;
+                Mod[] rulesetRequiredMods = requiredMods.Select(m => m.ToMod(userRuleset)).ToArray();
+
+                // In freestyle, the playlist item doesn't provide the allowed mods. Instead, all mods are unconditionally allowed by default.
+                return userRuleset.AllMods.OfType<Mod>()
```

Comment:
This cast is *very* dodgy, but given that removing it basically requires all helpers in this file to be converted to accept and return `IMod`, I'm not going to insist on changing anything. (That change should probably be done anyway at some point though.)